### ISSUE TITLE
Better Documentation

### ISSUE THREAD

**Issue by stckcrsh** [j-easy/easy-rules#164]
I have been trying to use the library, but i have had a few questions that i am trying to work through and this could easily be me not understanding this well enough.

but here goes

So i have some logic that has to check two parts of an object
Say we have a POJO like this
```
class TestPOJO {
  boolean legacy;
  List<int> array;
}
```

My logic needs to first see if this is legacy and if it is then the whole thing is valid.
If legacy is false then i need to validate that array has a size of one (invalid with any other size)

basically
```
if ( legacy == true ) { return true; }
if ( legacy == false) {
  if(array.size() == 1) {
    return true;
  } else {
    return false;
  }
}
```

My problem is that I want to do this using Rule Composition and i don't really see how i can get that done. I can easily just write all the logic into a single condition, but i really want to understand how to do logic like this using rule composition.

**Comment by anselm-klose** [comment]
I'm relativly new to EasyRules but I'll give it a try.
I assume you want to use `UnitRuleGroup` ?
You check on the [wiki page](https://github.com/j-easy/easy-rules/wiki/defining-rules).

Using `UnitRuleGroup` you can do this by creating two rules:

`@Rule(name = "rule1", description = "check if legacy is false", priority = 1)
public class MyRule {

    @Condition
    public boolean when(@Fact("legacy") Boolean legacy) {
        return !legacy;
    }

    @Action(order = 1)
    public void then(Facts facts) throws Exception {
    }
}`

`@Rule(name = "rule2", description = "check array size", priority = 1)
public class MyRule {

    @Condition
    public boolean when(@Fact("array") List<int> array) {
        //my rule conditions
        return array.size() == 1;
    }

    @Action(order = 1)
    public void then(Facts facts) throws Exception {
        facts.put("result", true)
    }
}

UnitRuleGroup myUnitRuleGroup =
    new UnitRuleGroup("myUnitRuleGroup", "unit of rule1 and rule2");
myUnitRuleGroup.addRule(myRule1);
myUnitRuleGroup.addRule(myRule2);
`

For all single line conditions just create separat rules.

**Comment by stckcrsh** [comment]
Sorry i am new to rule engines and i am trying to learn with this library.

@anselm-klose Thanks for the reply, I tried using the Unit Rule Group and it worked for most of my test cases. But I wrote out my logic a little wrong

I am trying to validate a request, the logic should be more like this
```
isValid(){
  if ( legacy == true ) { 
    return true; 
  } else {
    if(array.size() == 1) {
      return true;
    } else {
      return false;
    }
  }
}
```

This would be `OR` logic and i didn't see anything for doing logic that way.

I think the better question here is, am i even using the right tool for this job?

**Comment by stckcrsh** [comment]
@anselm-klose I completed it using a negative case which did it.
```
isNotValid(){
  return !legacy && array.size() != 1;
}
```

I think it would be nice though to have something like Condition composition instead of only rule composition.

I find that most of the condition logic is reusable throughout my app, but the actions are very distinct to the task at hand.  Which means i am frequently creating Rules with the RuleBuilder and not making actual rule classes.

**Comment by benas** [comment]
> I think it would be nice though to have something like Condition composition instead of only rule composition.

Condition composition is possible in both Java config and MVEL expressions:

Java config:

```java
class MyRule {

   boolean condition1() {}
   boolean condition2() {}

   @Condition
   public boolean condition() { return condition1() || condition2();} // for example

   // ... other rule parts
}
```

MVEL:

```yaml
-condition: !legacy && array.size() != 1;
```

However, you can't define multiple conditions in a rule descriptor file and try to combine them because Easy Rules can't statically interpret the combination as a 'OR' or a 'AND'. See https://github.com/j-easy/easy-rules/issues/193 for more details.

I'm closing this issue for now, but please don't hesitate to comment here if you need further help.
